### PR TITLE
 Create a test utility script that downloads test HTML and markdown and saves it as a part of the repo

### LINKED ISSUE #62: Create a test utility script that downloads test HTML and markdown and saves it as a part of the repo
Self assigned.

- Setup mocks of the `fetch()` method to use offline files. Leave a few copies of the test as integration tests calling actual Wiktionary, preferably shorter word pages that won't change any time soon.
- Create *scripts/download_wiktionary_test_files.py* that does the following:
  - Download HTML files of all words under test.
  - Download the Markdown files from the API endpoint below

```
https://en.wiktionary.org/w/api.php?action=parse&oldid={{OLDID}}&prop=wikitext&formatversion=2&format=json
```

### PR DESCRIPTION
**Pull request by sehwol**
Resolves #62

split-tests

The repo now has unit tests using offline copies and integration tests that call Wiktionary proper. All tests overall should now run faster.

The script for downloading test HTML and markup files is in the *scripts* folder. It does need `requests-futures==1.0.0` to run in parallel, which has been added to *requirements.txt*.

test-files

### CODE REVIEW

**Comment by Suyash458** [comment]
@sehwol this is great work! Do we need the markup files along with the HTML as well?

File: tests/test.py [hunk 1]
```diff
@@ -17,55 +78,16 @@ def __init__(self, *args, **kwargs):
...
-            self.__test_word(word, old_id, lang)
-
-    def __test_word(self, word: str, old_id: int, lang: str):
+    @parameterized.expand(get_test_words_table())
+    @mock.patch("requests.Session.get", side_effect=mocked_requests_get)
+    def test_fetch_using_mock_session(self, lang: str, word: str, old_id: int, mock_get):
+        self.__test_fetch(lang, word, old_id)
+
+    @parameterized.expand(get_test_words_table('ἀγγελία', 'test'))
+    def test_fetch_using_actual_session(self, lang: str, word: str, old_id: int):
```
Comment (Suyash458) [inline tests/test.py:87]: I'd say let's not make actual HTTP calls in the tests at all. We don't want to run into the same issue later where the page structure changes in the future and some of the tests start failing

**Comment by Suyash458** [comment]
I think we should also make some changes to `setup.py` so it packages the HTML files. The folder containing them will have to be added under `data_files`

**Comment by sehwol** [comment]
> 
> 
> Do we need the markup files along with the HTML as well?

My bad. I'll delete the markup files in this branch and comment it out in the utility script. Markup files will be necessary for testing #58 though, so it may reappear in that branch.

> 
> 
> I think we should also make some changes to `setup.py` so it packages the HTML files. The folder containing them will have to be added under `data_files`

Sorry, I don't use `data_files` that often, but can I ask the use-case for including *html_test_files* in the distribution? I'm just concerned it might increase the distribution size greatly if more HTML test files are added in the future.

*Image of the venv in one of my projects using WiktionaryParser*
venv

**Comment by sehwol** [comment]
Ok, deleted:
- Markup files.
- The related section in the util script.
- Actual HTTP calls in the tests.

**Comment by Suyash458** [comment]
>Sorry, I don't use data_files that often, but can I ask the use-case for including html_test_files in the distribution? I'm just concerned it might increase the distribution size greatly if more HTML test files are added in the future

Yeah it will add to the overall module size but, it will also make the repo self contained so the files needn't be download on every remote/local build. I'd added these specific words in the test because each of them had a slightly different HTML structure so, unless someone finds a page that's not covered by the existing ones, it needn't be added. 

The current size of the html test files is about 2MB which I think is fine. There are other repos (like [this](https://github.com/camelot-dev/camelot/tree/master/tests/files) one) as well that include large test resources in the repo

**Comment by sehwol** [comment]
Okay, I've added the HTML files using `glob`. I've tried installing it on another project and it seems to work as pictured below.

html_test_files_in_venv


And sorry to bother, but I'd like to ask/learn about `data_files` since I don't use Python at my workplace (only in personal projects).

If I include the test HTML files in the git repo but don't add them in *setup.py*, wouldn't remote/local builds of WiktionaryParser still work since the local computer/CI server `git clones` the whole repo? 

Or do you mean that other Python projects that rely on WiktionaryParser needs the HTML files for their own testing? The *setup.py* [file](https://github.com/camelot-dev/camelot/blob/master/setup.py) in the project you mentioned does not seem to use `data_files`, though the Git repo does include the [test files](https://github.com/camelot-dev/camelot/tree/master/tests/files) as you mentioned.

**Comment by Suyash458** [comment]
>If I include the test HTML files in the git repo but don't add them in setup.py, wouldn't remote/local builds of WiktionaryParser still work since the local computer/CI server git clones the whole repo? For example, the setup.py file in the project you mentioned does not seem to use data_files, though the Git repo does include the test files as you mentioned.

Yep you're right, CI builds would clone the repo and users who need to run tests would have to download the sources anyway. Adding it to `data_files` would unnecessarily include them in the build. Let's remove it from the `data_files` then. Sorry for the back and forth on this

> And sorry to bother, but I'd like to ask/learn about data_files since I don't use Python at my workplace aside from small automation scripts (only in personal projects).

No worries! The `data_files` configuration is used to package non code files along with the repo while publishing to pip. You may want to include certain configuration files or license files along with code.
https://python-packaging.readthedocs.io/en/latest/non-code-files.html

**Comment by sehwol** [comment]
Thank you for the explanation Suyash.

The `data_files` update has been reverted and the rebased for tidiness.

**Comment by Suyash458** [comment]
Thank you for adding this!

**Pull request merged**